### PR TITLE
Add support for flash-attn 3

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR integrates [flash-attn 3](https://github.com/Dao-AILab/flash-attention/hopper) into TE's `FlashAttention` module. This includes FP16/BF16 fwd+bwd, and FP8 fwd.

As of Aug 22, 2024, FA3 can be installed via the following commands:
```
1. pip install "git+https://github.com/Dao-AILab/flash-attention.git#egg=flashattn-hopper&subdirectory=hopper"
2. python_path=`python -c "import site; print(site.getsitepackages()[0])"`
3. mkdir -p $python_path/flashattn_hopper
4. wget -P $python_path/flashattn_hopper https://raw.githubusercontent.com/Dao-AILab/flash-attention/main/hopper/flash_attn_interface.py
```

Unit tests for FA3 can be found in `tests/pytorch/fused_attn/test_fused_attn.py`. For example, `test_dot_product_attention` for FP16/BF16, and `test_dpa_fp8_vs_f16` and `test_mha_fp8_vs_f16` for FP8.

When both FA3 (`flashattn-hopper`) and FA2 (`flash-attn`) are present in the environment, FA3 is prioritized - if it supports the specific user input. FA3 currently misses a few features such as ALiBi and sliding window, compared to FA2. To find out which FA version is being used, please run with `NVTE_DEBUG=1 NVTE_DEBUG_LEVEL=1`. For example,
```
# if FA3 is being used
[INFO     | DotProductAttention]: Running with FlashAttention backend (version 3.0.0b1)
# if FA2 is being used
[INFO     | DotProductAttention]: Running with FlashAttention backend (version 2.4.2)
```

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Add support for FA3

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by Eisenhower** [comment]
If FP8 is supported, does TE still need to be adapted separately?

File: tests/pytorch/fused_attn/test_fused_attn.py [hunk 1]
```diff
@@ -191,6 +191,12 @@ def test():
     "base_2_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "no_mask", "no_bias"),  # cross, 1
     "base_3_0": ModelConfig(8, 16, 16, 128, 1, 2048, 0.0, "no_mask", "no_bias"),  # inference
     "base_3_1": ModelConfig(8, 16, 16, 256, 1, 2048, 0.0, "no_mask", "no_bias"),  # inference
+    "base_11_0": ModelConfig(8, 16, 16, 64, 128, 128, 0.0, "causal", "no_bias"),  # self , 0
+    "base_11_1": ModelConfig(4, 16, 16, 64, 128, 256, 0.0, "causal", "no_bias"),  # cross, 0
+    "base_12_0": ModelConfig(2, 24, 24, 128, 2048, 2048, 0.0, "causal", "no_bias"),  # self , 1
+    "base_12_1": ModelConfig(1, 24, 24, 128, 2048, 4096, 0.0, "causal", "no_bias"),  # cross, 1
+    "base_13_0": ModelConfig(8, 16, 16, 128, 1, 2048, 0.0, "causal", "no_bias"),  # inference
+    "base_13_1": ModelConfig(8, 16, 16, 256, 1, 2048, 0.0, "causal", "no_bias"),  # inference
```
Comment (ksivaman) [inline tests/pytorch/fused_attn/test_fused_attn.py:199]: Not sure I understand the naming for these?

Comment (cyanguwa) [reply tests/pytorch/fused_attn/test_fused_attn.py]: It's removed, and some inference tests are added to `test_dpa_mask`.

**Comment by cyanguwa** [comment]
> If FP8 is supported, does TE still need to be adapted separately?

Yes, TE needs to update its `FlashAttention` module to use FA3's FP8 support.

**Comment by cyanguwa** [comment]
Pipeline 17767404 with custom FA3 installation.

**Comment by ksivaman** [comment]
@cyanguwa Is this ready for merge?

File: benchmarks/attention/benchmark_attention.py [hunk 2]
```diff
@@ -9,6 +9,8 @@
 import torch
 import nvtx
 import transformer_engine
+
+sys.path.append("../../")
```
Comment (ksivaman) [inline benchmarks/attention/benchmark_attention.py:13]: Hmm, even though this is a standalone benchmark, this isn't ideal

Comment (cyanguwa) [reply benchmarks/attention/benchmark_attention.py]: Removed. There were some changes I made when debugging stuff.

**Comment by cyanguwa** [comment]
> @cyanguwa Is this ready for merge?

Not yet. I'm doing some CI.

**Review by ksivaman — APPROVED** [review]
LGTM!

**Pull request merged**